### ISSUE TITLE
Switching back to application with no open windows

### ISSUE THREAD

**Issue by a3b4c6** [lwouis/alt-tab-macos#664]
**Is your feature suggestion related to a problem? Please describe.**
After switching *away* from an application with no windows, I cannot quickly switch back to the application again. An example: if Safari is the frontmost application and it has no open windows, then after switching to another application, I can no longer alt-tab back to Safari with a single keystroke. In the current version of alt-tab-macos, all applications without windows are grouped together after all open windows.

**Describe the solution you'd like**
I propose adding two options:
- Add the option to group *both* (1) all open windows and (2) all applications without any windows together in chronological order instead of having all of (1) before all of (2).
- Add an option so that switching to an application with no open windows simply focuses the application without creating any new windows.

**Describe alternatives you've considered**
Use macos' native command-tab instead of alt-tab-macos in this situation. This is unsatisfactory because alt-tab-macos then cannot replace macos' command-tab functionality.

**Comment by lwouis** [comment]
Hi @a3b4c6! Thanks for sharing this suggestion!

I'm curious to understand the use-case better. You said for instance you want to switch back to Safari, but not open a new window? (the second option you listed). Why would you not want a new window in that scenario? What are you going to do with Safari without opening a window?

Regarding the first option you listed, we could move the preferences from a checkbox:

![image](https://user-images.githubusercontent.com/106195/97206006-89bc9600-17fb-11eb-9581-843c05772ca6.png)

to a dropdown menu like this one:

![image](https://user-images.githubusercontent.com/106195/97205981-81fcf180-17fb-11eb-852e-ad3a733721de.png)

I wonder if it should be per-shortcut though, or global. I think most people would prefer global?

**Comment by a3b4c6** [comment]
Thanks you for looking into this. I will try to elaborate on my original suggestion.

> I'm curious to understand the use-case better. You said for instance you want to switch back to Safari, but not open a new window? (the second option you listed). Why would you not want a new window in that scenario? What are you going to do with Safari without opening a window?

Here is a more detailed case for the second option: Not opening a new window when activating an application is the behavior I presume many macos users will be familiar with from the built-in <kbd>cmd+tab</kbd>. Since an application's shortcuts and menu bar are enabled even when there is no window open, there are often many ways to interact with an application without opening a window.

Here are some examples:
- I have an image in the clipboard and I want to create a new document in preview using `File -> New from Clipboard`. Creating a new window, i.e. displaying an "Open File" dialog is not necessary.
- I want to focus an application and just open the Preferences window. Creating a new browser window or document is not necessary for that purpose either.
- Not all applications have "create new document" or equivalent as their default "new window" behaviour. Typora for example currently opens the least recently closed document when you focus it in AltTab.
- There can be more than one way of creating a new main window, e.g. `File -> New Window` and `File -> New Private Window` in Firefox.

Not only is creating a window unnecessary in these situations (and in many similar cases), but it can be jarring for a user to have a window pop up when the intention is to do something else. This is especially true for applications that are less snappy, which have an obvious delay between activating and actually loading a new window.

> Regarding the first option you listed, we could move the preferences from a checkbox

That would indeed work for my original suggestion. However, after thinking a bit about what I want as a user, I'm not completely sure what I suggested would be sufficient, at least not for AltTab to be a replacement for <kbd>cmd+tab</kbd>.

At the moment, I use Shortcut 1 in AltTab to only show windows from the current space, and Shortcut 2 to show windows from all spaces. For both Shortcut 1 and Shortcut 2, AltTab currently lists all window-less applications at the end of the switcher.

I think what I really need is not just for the window-less applications to be in the switcher, but all running apps that do not have a window in the current space. That would make the functionarlity of AltTab a superset of <kbd>cmd+tab</kbd>.

Say I have a Safari window in Space 1 and I'm currently in Space 2. How do I create a new Safari window in the current space using AltTab? If it is possible to switch to Safari without activating the window in Space 1, I can create a new window in the current space by hitting <kbd>cmd+n</kbd>.

I'm not completely sure, but I think it would be best if the windows and applications are displayed in chronological order, even if the most recent activation of e.g. Safari was in another space.

**Comment by lwouis** [comment]
I would like to clarify that though AltTab shows an icon of a window with a plus sign, and if you focus it, you usually focus that app + indeed create a new window, it is not actually the literal implementation. When implementing #397, I identified 2 strategies:

* Imitate macOS app switcher, and simply focus the app [1]
* Imitate a double-click on the app icon, which is the same as a single-click on the app icon in the Dock

I went with the second option, which means AltTab is in fact "launching" the app. There is no way for AltTab to tell an app to open a window. By launching it, however, it sends it the signal from the interactions I listed above, and most apps have an appropriate reaction to that signal. Apps who don't implement it just get focused I believe.

Anyway, we could add an option to simply focus the app instead of launching it. I'm wondering how many people would want that though. You're the first person to desire such a workflow. I don't want to sound like I have an agenda for the current way of doing things. I'm just hearing this request for the first time in this ticket, so I'm wondering if it's a niche workflow, and if more users would desire that.

Regarding the chronological order, I think it's a clear use-case as most of the usefulness of AltTab is chronologically navigate your windows/apps. I was asked to implement the current way during the original implementation in #397, but we can add a preference today I think.

[1] I think this behavior actually confuses many users. I've seen many friends say like "Where is Safari?" after they switch to it. I would then point to the narrow bar at the top and say "See, it's active right now". I think it's confusing UX to be honest, and overall I think I think the Windows model of "close last window = close the app, except some apps that "go into the taskbar as a small icon"" is clearer. It could have been implemented to macOS with most apps just closing when the last window is closed, and a few appearing in the top menubar in such occasions to show they are always opened basically.
